executor/new_index_lookup_join.go

```diff
@@ -0,0 +1,531 @@
+		return datumRow, nil
+	}
+}
+
+func (e *NewIndexLookUpJoin) prepareJoinResult(goCtx goctx.Context) error {
+	if e.joinResultCursor < e.joinResult.NumRows() {
+		return nil
+	}
+	e.joinResult.Reset()
+	e.joinResultCursor = 0
+	for {
+		task, err := e.getFinishedTask(goCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if task == nil {
+			return nil
+		}
+		for task.cursor < task.outerResult.NumRows() {
+			e.lookUpMatchedInners(task, task.cursor)
+			outerRow := task.outerResult.GetRow(task.cursor)
+			task.cursor++
+			err = e.resultGenerator.emitToChunk(outerRow, task.matchedInners, e.joinResult)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if e.joinResult.NumRows() > 0 {
+				return nil
+			}
+		}
+	}
+}
+
+func (e *NewIndexLookUpJoin) getFinishedTask(goCtx goctx.Context) (*lookUpJoinTask, error) {
+	task := e.task
+	if task != nil && task.cursor < task.outerResult.NumRows() {
+		return task, nil
+	}
+	select {
+	case task = <-e.resultCh:
+	case <-goCtx.Done():
+		return nil, nil
+	}
+	if task == nil {
+		return nil, nil
+	}
+
+	select {
+	case err := <-task.doneCh:
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	case <-goCtx.Done():
+		return nil, nil
+	}
+	e.task = task
+	return task, nil
+}
+
+func (e *NewIndexLookUpJoin) lookUpMatchedInners(task *lookUpJoinTask, rowIdx int) {
+	outerKey := task.encodedLookUpKeys.GetRow(rowIdx).GetBytes(0)
+	innerPtrBytes := task.lookupMap.Get(outerKey)
+	task.matchedInners = task.matchedInners[:0]
+	for _, b := range innerPtrBytes {
+		ptr := *(*chunk.RowPtr)(unsafe.Pointer(&b[0]))
+		matchedInner := task.innerResult.GetRow(ptr)
+		task.matchedInners = append(task.matchedInners, matchedInner)
+	}
+}
+
+func (ow *outerWorker) run(goCtx goctx.Context, wg *sync.WaitGroup) {
+	defer func() {
+		close(ow.resultCh)
+		close(ow.innerCh)
+		wg.Done()
+	}()
+	for {
+		task, err := ow.buildTask(goCtx)
+		if err != nil {
+			task.doneCh <- errors.Trace(err)
+			return
+		}
+		if task == nil {
+			return
+		}
+		select {
+		case <-goCtx.Done():
+			return
+		case ow.resultCh <- task:
+		}
+		select {
+		case <-goCtx.Done():
+			return
+		case ow.innerCh <- task:
+		}
+	}
+}
+
+// buildTask builds a lookUpJoinTask and read outer rows.
+// When err is not nil, task must not be nil to send the error to the main thread via task.
+func (ow *outerWorker) buildTask(goCtx goctx.Context) (*lookUpJoinTask, error) {
+	ow.executor.newChunk()
+	task := new(lookUpJoinTask)
+	task.doneCh = make(chan error, 1)
+	task.outerResult = ow.executor.newChunk()
+	task.encodedLookUpKeys = chunk.NewChunk([]*types.FieldType{types.NewFieldType(mysql.TypeBlob)})
+	task.lookupMap = mvmap.NewMVMap()
+	ow.increaseBatchSize()
+
+	for task.outerResult.NumRows() < ow.batchSize {
+		err := ow.executor.NextChunk(goCtx, ow.executorChk)
+		if err != nil {
+			return task, errors.Trace(err)
+		}
+		if ow.executorChk.NumRows() == 0 {
+			break
+		}
+		task.outerResult.Append(ow.executorChk, 0, ow.executorChk.NumRows())
```

Comment:
after this, `task.outerResult.NumRows()` can be larger than `ow.batchSize`

Reply:
I know, but it's simpler.